### ISSUE TITLE
Msal caches multiple access tokens for the same resource, retrieves incorrect access token from the cache

### ISSUE THREAD

**Issue by salman90** [AzureAD/microsoft-authentication-library-for-js#4523]
### Core Library

MSAL.js v2 (@azure/msal-browser)

### Core Library Version

2.16.0

### Wrapper Library

MSAL React (@azure/msal-react)

### Wrapper Library Version

1.3.0

### Description

I have a Node web API that rejects any access token coming from the frontend that doesn’t contain a certain conditional access claim. When this happens, I make a new token request on my frontend, passing the claims that I need in my request object  (i.e. claims challenge)

```javascript
const handleClaimsChallenge = async (response) => {
    if (response.status === 401) {
        if (response.headers.get('www-authenticate')) {
            const authenticateHeader = response.headers.get("www-authenticate");

            const claimsChallenge = authenticateHeader.split(" ")
                .find(entry => entry.includes("claims=")).split('="')[1].split('",')[0];

            try {
                await msalInstance.acquireTokenPopup({
                    claims: window.atob(claimsChallenge), // decode the base64 string
                            scopes: ["api://xxx-xxxx/access_as_user"], 

                });
            } catch (error) {
                // catch if popups are blocked
                if (error instanceof BrowserAuthError && 
                    (error.errorCode === "popup_window_error" || error.errorCode === "empty_window_error")) {

                    await msalInstance.acquireTokenRedirect({
                        claims: window.atob(claimsChallenge),
                        scopes: protectedResources.apiTodoList.scopes
                    });
                }
            }
        } else {
            return { error: "unknown header" }
        }
    }

    return response.json();
}
```
However, msal caches this new access-token separately. This leads to multiple access-tokens for the same resource stored in the browser cache with different keys:
<img width="727" alt="tokens" src="https://user-images.githubusercontent.com/18670157/154085594-83ca563f-39d8-4531-8282-77373475f39b.png">

When acquireTokenSilent fetches the access-token from the cache, it goes for the first access token it finds in the cache, which does not include the relevant claim that was requested via claims challenge.

```javascript
const getToken = async () => {
    const account = msalInstance.getActiveAccount();
    if (!account) {
        throw Error("No active account! Verify a user has been signed in and setActiveAccount has been called.");
    }

    const response = await msalInstance.acquireTokenSilent({
        account: account,
        scopes: ["api://xxx-xxx/access_as_user"]
    });
}
```

### MSAL Configuration

```javascript
import { LogLevel } from "@azure/msal-browser";

/**
 * Configuration object to be passed to MSAL instance on creation. 
 * For a full list of MSAL.js configuration parameters, visit:
 * https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-browser/docs/configuration.md 
 */
export const msalConfig = {
    auth: {
        clientId: 'clientId', // This is the ONLY mandatory field that you need to supply.
        authority: 'https://login.microsoftonline.com/tenantId', // Defaults to "https://login.microsoftonline.com/common"
        redirectUri: "http://localhost:3000", // You must register this URI on Azure Portal/App Registration. Defaults to window.location.origin
        postLogoutRedirectUri: "http://localhost:3000", // Indicates the page to navigate after logout.
        navigateToLoginRequestUrl: true, // If "true", will navigate back to the original request location before processing the auth code response.
        clientCapabilities: ["CP1"] // this lets the resource owner know that this client is capable of handling claims challenge.
    },
    cache: {
        cacheLocation: "localStorage",
        storeAuthStateInCookie: false, 
    },
    system: {
        loggerOptions: {
            loggerCallback: (level, message, containsPii) => {
                if (containsPii) {
                    return;
                }
                switch (level) {
                    case LogLevel.Error:
                        console.error(message);
                        return;
                    case LogLevel.Info:
                        console.info(message);
                        return;
                    case LogLevel.Verbose:
                        console.debug(message);
                        return;
                    case LogLevel.Warning:
                        console.warn(message);
                        return;
                }
            }
        }
    }
};

export const protectedResources = {
    apiTodoList: {
        todoListEndpoint: "http://localhost:5000/api/todolist",
        scopes: [" api://xxxxx-xxxx-xxxx-xxxx-xxxxx/access_as_user"],
    },
}

export const loginRequest = {
    scopes: [...protectedResources.apiTodoList.scopes],
};


// .env file

AUTHORITY="login.microsoftonline.com"

TENANT_ID="TENANT_ID"
CLIENT_ID="CLIENT_ID"
CLIENT_SECRET="CLIENT_SECRET"

REDIRECT_URI="http://localhost:5000/admin/redirect"

API_REQUIRED_PERMISSION="access_as_user"

EXPRESS_SESSION_SECRET="ENTER_YOUR_SECRET_HERE"

CORS_ALLOWED_DOMAINS="http://localhost:3000"
```

### Relevant Code Snippets

```javascript
const handleClaimsChallenge = async (response) => {
    if (response.status === 401) {
        if (response.headers.get('www-authenticate')) {
            const authenticateHeader = response.headers.get("www-authenticate");

            const claimsChallenge = authenticateHeader.split(" ")
                .find(entry => entry.includes("claims=")).split('="')[1].split('",')[0];

            try {
                await msalInstance.acquireTokenPopup({
                    claims: window.atob(claimsChallenge), // decode the base64 string
                            scopes: ["api://xxx-xxxx/access_as_user"], 

                });
            } catch (error) {
                // catch if popups are blocked
                if (error instanceof BrowserAuthError && 
                    (error.errorCode === "popup_window_error" || error.errorCode === "empty_window_error")) {

                    await msalInstance.acquireTokenRedirect({
                        claims: window.atob(claimsChallenge),
                        scopes: protectedResources.apiTodoList.scopes
                    });
                }
            }
        } else {
            return { error: "unknown header" }
        }
    }

    return response.json();
```
```javascript
const handleClaimsChallenge = async (response) => {
    if (response.status === 401) {
        if (response.headers.get('www-authenticate')) {
            const authenticateHeader = response.headers.get("www-authenticate");

            const claimsChallenge = authenticateHeader.split(" ")
                .find(entry => entry.includes("claims=")).split('="')[1].split('",')[0];

            try {
                await msalInstance.acquireTokenPopup({
                    claims: window.atob(claimsChallenge), // decode the base64 string
                            scopes: ["api://xxx-xxxx/access_as_user"], 

                });
            } catch (error) {
                // catch if popups are blocked
                if (error instanceof BrowserAuthError && 
                    (error.errorCode === "popup_window_error" || error.errorCode === "empty_window_error")) {

                    await msalInstance.acquireTokenRedirect({
                        claims: window.atob(claimsChallenge),
                        scopes: protectedResources.apiTodoList.scopes
                    });
                }
            }
        } else {
            return { error: "unknown header" }
        }
    }

    return response.json();
}
```

```javascript
//get token
const getToken = async () => {
    const account = msalInstance.getActiveAccount();
    if (!account) {
        throw Error("No active account! Verify a user has been signed in and setActiveAccount has been called.");
    }

    const response = await msalInstance.acquireTokenSilent({
        account: account,
        scopes: ["api://xxx-xxx/access_as_user"]
    });
}
```

### Reproduction Steps

1- clone sample [Use Conditional Access Authentication Context](https://github.com/Azure-Samples/ms-identity-javascript-react-tutorial/blob/main/6-AdvancedScenarios/3-call-api-acrs/README.md)
2- update msal-react

### Expected Behavior

Fetching the access-token with correct access claims 

### Identity Provider

Azure AD / MSA

### Browsers Affected (Select all that apply)

Chrome

### Regression

 "@azure/msal-react": "^1.0.1"

### Source

Internal (Microsoft)

**Comment by hectormmg** [comment]
This seems to be related to a bug introduced in v 2.21 with regards to matching the incorrect access token when doing a cache lookup for a token that includes client claims. Looking into it, will link the PR with the fix soon.

**Comment by salman90** [comment]
@hectormmg thanks for the update.

**Comment by hectormmg** [comment]
After reviewing this issue and discussing it offline with @salman90 we determined this isn't a bug, but the intended behavior. The actual bug around returning the incorrect token when claims are requested is only present in the current version of MSAL Node.

What is happening here is that there are two tokens in the cache for the same resource, one with and one without claims. These aren't duplicated, since the claims differentiate them. In order to obtain the token with the claims, the claims must be added to the request every time, not only when going to the network.

Closing since the issue has been clarified, we can reopen if necessary.
